File: examples and misc/Java/Abstraction.java
public class Abstraction {
    // Abstracts class
    abstract static class Animal {
        abstract void speak(); // Abstract method
    }

    // Derived class
    static class Dog extends Animal {
        @Override
        void speak() {
            System.out.println("Dog barks.");
        }
    }

    // Derived class
    static class Cat extends Animal {
        @Override
        void speak() {
            System.out.println("Cat meows.");
        }
    }

    public static void main(String[] args) {
        Animal myDog = new Dog();
        Animal myCat = new Cat();
        myDog.speak();
        myCat.speak();
    }
}

File: examples and misc/Cpp/abstraction.cpp
#include <iostream>
using namespace std;

// Abstract class
class Animal {
public:
    virtual void speak() = 0; // Pure virtual function
};

// Derived class
class Dog : public Animal {
public:
    void speak() override {
        cout << "Dog barks." << endl;
    }
};

// Derived class
class Cat : public Animal {
public:
    void speak() override {
        cout << "Cat meows." << endl;
    }
};

int main() {
    Animal* myDog = new Dog();
    Animal* myCat = new Cat();
    myDog->speak();
    myCat->speak();
    return 0;
}
